src/envoy/tcp/mixer/control.cc

```diff
@@ -25,21 +25,20 @@ namespace Envoy {
 namespace Tcp {
 namespace Mixer {
 
-Control::Control(const Config& config, Upstream::ClusterManager& cm,
-                 Event::Dispatcher& dispatcher,
+Control::Control(ControlDataSharedPtr control_data,
```

Comment:
Why pass in shared_ptr?   If you don't want "Control" to increase ref,  don't pass in shared_ptr.   Just pass in "const &"

Reply:
Changed to ControlData& control_data. We cannot pass const ControlData& because control_data.stats() could not become const.

Reply:
In this case, couldn't config_ be a dangling pointer when ControlFactory is gone?

Reply:
I am not clear with this comment. Are you saying that passing ControlData& control_data is not safe, control_data is destroyed when ControlFactory is gone? We pass ControlDataSharedPtr control_data_ into capture list, that increases the ref count, no?

Reply:
ThreadLocalObject::set does't hold the lambda, it just post to all worker thread. So once the execution is done, the only ref of ControlDataSharedPtr is ControlFactory, and since Control holds config as const reference, it might be dangling if ControlFactory is gone.

Reply:
The question is: will Control (the object stored in each thread local slot) out-live ControlFactory.    Or another word, Will the object stored in tls_ out live tls_ itself?  I guess the answer is yes, the it is better to pass the ControlDataSharedPtr to Control and Control has to hold a reference count

Reply:
> object stored in tls_ out live tls_ itself

Yes, slot destroy all data by posting callback to all dispatchers.

Reply:
I add a member ControlDataSharedPtr control_data_ into Http::Mixer::Control and Tcp::Mixer::Control, and add a JwtAuthenticationConstSharedPtr config_ into JwtAuthStore. This is to guarantee that each worker thread hold a ref to the shared config, and the config ref count is always positive. So that even ControlFactory or JwtAuthStoreFactore is gone, JwtAuthStore and Control objects do not cause envoy to crash.

Reply:
Please take a look. I have built and run binary in istio.io/istio and all tests pass.

Reply:
Thanks for manually testing against istio/istio.